File: killrvideo-service-videocatalog/src/test/java/com/killrvideo/service/video/grpc/VideoCatalogServiceGrpcMapperTest.java
package com.killrvideo.service.video.grpc;

import com.killrvideo.dse.dto.CustomPagingState;
import com.killrvideo.dse.dto.ResultListPage;
import com.killrvideo.dse.dto.Video;
import com.killrvideo.service.video.dto.LatestVideo;
import com.killrvideo.service.video.dto.LatestVideosPage;
import com.killrvideo.service.video.dto.UserVideo;
import com.killrvideo.service.video.request.GetLatestVideoPreviewsRequestData;
import com.killrvideo.service.video.request.GetUserVideoPreviewsRequestData;
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.*;
import killrvideo.video_catalog.events.VideoCatalogEvents;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.HashSet;
import java.util.Optional;
import java.util.UUID;

import static com.killrvideo.utils.GrpcMappingUtils.instantToTimeStamp;
import static com.killrvideo.utils.GrpcMappingUtils.uuidToUuid;
import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

class VideoCatalogServiceGrpcMapperTest {
    private final VideoCatalogServiceGrpcMapper mapper = new VideoCatalogServiceGrpcMapper();

    @Test
    public void testMapSubmitYouTubeVideoRequestAsVideo() {
        Video v = video();
        SubmitYouTubeVideoRequest request = SubmitYouTubeVideoRequest.newBuilder()
                .setVideoId(uuidToUuid(v.getVideoid()))
                .setUserId(uuidToUuid(v.getUserid()))
                .setName(v.getName())
                .setDescription(v.getDescription())
                .setYouTubeVideoId(v.getLocation())
                .addAllTags(v.getTags())
                .build();

        Video video = mapper.mapSubmitYouTubeVideoRequestAsVideo(request);
        assertEquals(v.getVideoid(), video.getVideoid());
        assertEquals(v.getUserid(), video.getUserid());
        assertEquals(v.getName(), video.getName());
        assertEquals(v.getLocation(), video.getLocation());
        assertEquals(v.getDescription(), video.getDescription());
        assertEquals(v.getTags().size(), video.getTags().size());
        assertEquals(v.getLocationType(), video.getLocationType());
    }

    @Test
    public void testMapLatestVideoToGrpcResponse() {
        Video v = video();
        LatestVideo latestVideo = LatestVideo.from(v, Instant.now());
        String nextPageState = "next page state";
        String cassandraPagingState = "cassandra paging state";
        LatestVideosPage latestVideosPage = new LatestVideosPage(singletonList(latestVideo),
                cassandraPagingState,
                nextPageState);

        GetLatestVideoPreviewsResponse response = mapper.mapLatestVideoToGrpcResponse(latestVideosPage);
        assertEquals(nextPageState, response.getPagingState());
        assertEquals(1, response.getVideoPreviewsCount());
    }

    @Test
    public void testMapFromVideotoVideoResponse() {
        Video v = video();
        GetVideoResponse response = mapper.mapFromVideotoVideoResponse(v);

        assertEquals(v.getVideoid().toString(), response.getVideoId().getValue());
        assertEquals(v.getDescription(), response.getDescription());
        assertEquals(v.getLocation(), response.getLocation());
        assertEquals(v.getLocationType(), response.getLocationType().getNumber());
        assertEquals(v.getName(), response.getName());
        assertEquals(v.getUserid().toString(), response.getUserId().getValue());
        assertEquals(v.getTags().size(), response.getTagsCount());
    }

    @Test
    public void testCreateYouTubeVideoAddedEvent() {
        Video v = video();
        VideoCatalogEvents.YouTubeVideoAdded event = mapper.createYouTubeVideoAddedEvent(v);

        assertEquals(v.getVideoid().toString(), event.getVideoId().getValue());
        assertEquals(v.getDescription(), event.getDescription());
        assertEquals(v.getLocation(), event.getLocation());
        assertEquals(v.getName(), event.getName());
        assertEquals(v.getUserid().toString(), event.getUserId().getValue());
    }

    @Test
    public void testParseGetLatestVideoPreviewsRequest() {
        UUID startingVidoId = UUID.randomUUID();
        GetLatestVideoPreviewsRequest request = GetLatestVideoPreviewsRequest.newBuilder()
                .setPageSize(2)
                .setStartingVideoId(uuidToUuid(startingVidoId))
                .build();

        GetLatestVideoPreviewsRequestData pojo = mapper.parseGetLatestVideoPreviewsRequest(
                request,
                CustomPagingState::buildFirstCustomPagingState
        );

        assertEquals(2, pojo.getPageSize());
        assertNotNull(pojo.getPageState());
        assertEquals(startingVidoId, pojo.getStartVideoId().get());
    }

    @Test
    public void testMapToGetVideoPreviewsResponse() {
        Video v = video();
        GetVideoPreviewsResponse response = mapper.mapToGetVideoPreviewsResponse(singletonList(v));

        assertEquals(1, response.getVideoPreviewsCount());
        VideoPreview preview = response.getVideoPreviews(0);

        assertEquals(v.getName(), preview.getName());
        assertEquals(v.getUserid().toString(), preview.getUserId().getValue());
        assertEquals(v.getVideoid().toString(), preview.getVideoId().getValue());
    }

    @Test
    public void testParseGetUserVideoPreviewsRequest() {
        UUID userid = UUID.randomUUID();
        UUID startingVideioId = UUID.randomUUID();
        Instant startingAddedDate = Instant.now().minus(1L, ChronoUnit.DAYS);
        int pageSize = 2;
        String pagingState = "Paging state";

        GetUserVideoPreviewsRequest request = GetUserVideoPreviewsRequest.newBuilder()
                .setUserId(uuidToUuid(userid))
                .setStartingVideoId(uuidToUuid(startingVideioId))
                .setStartingAddedDate(instantToTimeStamp(startingAddedDate))
                .setPageSize(pageSize)
                .setPagingState(pagingState)
                .build();

        GetUserVideoPreviewsRequestData pojo = mapper.parseGetUserVideoPreviewsRequest(request);

        assertEquals(userid, pojo.getUserId());
        assertEquals(startingVideioId, pojo.getStartingVideoId().get());
        assertEquals(startingAddedDate, pojo.getStartingAddedDate().get());
        assertEquals(pageSize, pojo.getPagingSize().get());
        assertEquals(pagingState, pojo.getPagingState().get());
    }

    @Test
    public void testMapToGetUserVideoPreviewsResponse() {
        UUID userid = UUID.randomUUID();
        Video v = video();
        UserVideo userVideo = UserVideo.from(v, Instant.now());
        String pagingState = "Paging state";
        ResultListPage<UserVideo> resultListPage = new ResultListPage<>(
                singletonList(userVideo),
                Optional.of(pagingState)
        );

        GetUserVideoPreviewsResponse response = mapper.mapToGetUserVideoPreviewsResponse(resultListPage, userid);
        assertEquals(pagingState, response.getPagingState());
        assertEquals(1, response.getVideoPreviewsCount());
    }

    private Video video() {
        Video v = new Video();
        v.setVideoid(UUID.randomUUID());
        v.setUserid(UUID.randomUUID());
        v.setName("Game");
        v.setLocation("url");
        v.setDescription("Description");
        v.setTags(new HashSet<>(asList("tag1", "tag2")));
        v.setPreviewImageLocation("previewUrl");
        v.setLocationType(VideoLocationType.YOUTUBE.ordinal());
        v.setAddedDate(Instant.now());
        return v;
    }
}

File: killrvideo-service-videocatalog/src/test/kotlin/com/killrvideo/service/video/grpc/VideoCatalogServiceGrpcMapperTest.kt
package com.killrvideo.service.video.grpc

import com.killrvideo.dse.dto.CustomPagingState
import com.killrvideo.dse.dto.ResultListPage
import com.killrvideo.service.video.dto.LatestVideo
import com.killrvideo.service.video.dto.LatestVideosPage
import com.killrvideo.service.video.dto.UserVideo
import com.killrvideo.service.video.dto.Video
import com.killrvideo.service.video.grpc.VideoCatalogServiceGrpcMapper.GetLatestVideoPreviewsRequestExtensions.parse
import com.killrvideo.service.video.grpc.VideoCatalogServiceGrpcMapper.GetUserVideoPreviewsRequestExtensions.parse
import com.killrvideo.service.video.grpc.VideoCatalogServiceGrpcMapper.SubmitYouTubeVideoRequestExtensions.parse
import com.killrvideo.utils.GrpcMappingUtils.instantToTimeStamp
import com.killrvideo.utils.GrpcMappingUtils.uuidToUuid
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.VideoLocationType
import killrvideo.video_catalog.getLatestVideoPreviewsRequest
import killrvideo.video_catalog.getUserVideoPreviewsRequest
import killrvideo.video_catalog.submitYouTubeVideoRequest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*

internal class VideoCatalogServiceGrpcMapperTest {
    private val mapper = VideoCatalogServiceGrpcMapper()

    @Test
    fun testMapSubmitYouTubeVideoRequestAsVideo() {
        val v = video()
        val request = submitYouTubeVideoRequest {
            videoId = uuidToUuid(v.videoid)
            userId = uuidToUuid(v.userid)
            name = v.name!!
            description = v.description!!
            youTubeVideoId = v.location!!
            tags.addAll(v.tags!!)
        }
        val video = request.parse()
        assertEquals(v.videoid, video.videoid)
        assertEquals(v.userid, video.userid)
        assertEquals(v.name, video.name)
        assertEquals(v.location, video.location)
        assertEquals(v.description, video.description)
        assertEquals(v.tags!!.size, video.tags!!.size)
        assertEquals(v.locationType, video.locationType)
    }

    @Test
    fun testMapLatestVideoToGrpcResponse() {
        val v = video()
        val latestVideo = LatestVideo.from(v, Instant.now())
        val nextPageState = "next page state"
        val cassandraPagingState = "cassandra paging state"
        val latestVideosPage = LatestVideosPage(
            listOfPreview = mutableListOf(latestVideo),
            cassandraPagingState = cassandraPagingState,
            nextPageState = nextPageState
        )
        val response = mapper.mapLatestVideoToGrpcResponse(latestVideosPage)
        assertEquals(nextPageState, response.pagingState)
        assertEquals(1, response.videoPreviewsCount)
    }

    @Test
    fun testMapFromVideotoVideoResponse() {
        val v = video()
        val response = mapper.mapFromVideotoVideoResponse(v)
        assertEquals(v.videoid.toString(), response.videoId.value)
        assertEquals(v.description, response.description)
        assertEquals(v.location, response.location)
        assertEquals(v.locationType, response.locationType.number)
        assertEquals(v.name, response.name)
        assertEquals(v.userid.toString(), response.userId.value)
        assertEquals(v.tags!!.size, response.tagsCount)
    }

    @Test
    fun testCreateYouTubeVideoAddedEvent() {
        val v = video()
        val event = mapper.createYouTubeVideoAddedEvent(v)
        assertEquals(v.videoid.toString(), event.videoId.value)
        assertEquals(v.description, event.description)
        assertEquals(v.location, event.location)
        assertEquals(v.name, event.name)
        assertEquals(v.userid.toString(), event.userId.value)
    }

    @Test
    fun testParseGetLatestVideoPreviewsRequest() {
        val startingVidoId = UUID.randomUUID()
        val request = getLatestVideoPreviewsRequest {
            pageSize = 2
            startingVideoId = uuidToUuid(startingVidoId)
        }
        val pojo = request.parse { CustomPagingState.buildFirstCustomPagingState() }
        assertEquals(2, pojo.pageSize)
        assertNotNull(pojo.startVideoId)
        assertEquals(startingVidoId, pojo.startVideoId)
    }

    @Test
    fun testMapToGetVideoPreviewsResponse() {
        val v = video()
        val response = mapper.mapToGetVideoPreviewsResponse(listOf(v))
        assertEquals(1, response.videoPreviewsCount)
        val preview = response.getVideoPreviews(0)
        assertEquals(v.name, preview.name)
        assertEquals(v.userid.toString(), preview.userId.value)
        assertEquals(v.videoid.toString(), preview.videoId.value)
    }

    @Test
    fun testParseGetUserVideoPreviewsRequest() {
        val userid = UUID.randomUUID()
        val startingVideoid = UUID.randomUUID()
        val startingDate = Instant.now().minus(1L, ChronoUnit.DAYS)
        val pagesize = 2
        val pagingstate = "Paging state"
        val request = getUserVideoPreviewsRequest {
            userId = uuidToUuid(userid)
            startingVideoId = uuidToUuid(startingVideoid)
            startingAddedDate = instantToTimeStamp(startingDate)
            pageSize = pagesize
            pagingState = pagingstate
        }
        val pojo = request.parse()
        assertEquals(userid, pojo.userId)
        assertNotNull(pojo.startingVideoId)
        assertEquals(startingVideoid, pojo.startingVideoId)
        assertNotNull(pojo.startingAddedDate)
        assertEquals(startingDate, pojo.startingAddedDate)
        assertNotNull(pojo.pagingSize)
        assertEquals(pagesize, pojo.pagingSize)
        assertNotNull(pojo.pagingState)
        assertEquals(pagingstate, pojo.pagingState)
    }

    @Test
    fun testMapToGetUserVideoPreviewsResponse() {
        val userid = UUID.randomUUID()
        val v = video()
        val userVideo = UserVideo.from(v, Instant.now())
        val pagingState = "Paging state"
        val resultListPage = ResultListPage(
            listOf(userVideo),
            Optional.of(pagingState)
        )
        val response = mapper.mapToGetUserVideoPreviewsResponse(resultListPage, userid)
        assertEquals(pagingState, response.pagingState)
        assertEquals(1, response.videoPreviewsCount)
    }

    private fun video(): Video =
        Video(
            videoid = UUID.randomUUID(),
            userid = UUID.randomUUID(),
            name = "Game",
            location = "url",
            description = "Description",
            tags = mutableSetOf("tag1", "tag2"),
            previewImageLocation = "previewUrl",
            locationType = VideoLocationType.YOUTUBE.ordinal,
            addedDate = Instant.now()
        )
}
